File: src/data_retriever.py
"""
Historical data retriever for cryptocurrency data from Coinbase API.
Handles data fetching, validation, transformation, and retry logic.
"""

from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any
from decimal import Decimal
import structlog
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type

from coinbase_client import coinbase_client
from models import CryptoPriceData, DataRetrievalRequest, DataRetrievalResult, SymbolValidator
from database import DatabaseManager

logger = structlog.get_logger(__name__)


class HistoricalDataRetriever:
    """Fetches historical cryptocurrency data from Coinbase API."""
    
    def __init__(self):
        """Initialize historical data retriever."""
        self.client = coinbase_client
        self._db_managers = {}  # Cache for granularity-specific database managers
        
        if not self.client.is_authenticated:
            logger.error("Coinbase client not authenticated")
            raise ValueError("Coinbase client authentication required")
    
    def get_database_manager(self, granularity: int) -> DatabaseManager:
        """Get database manager for specific granularity."""
        if granularity not in self._db_managers:
            self._db_managers[granularity] = DatabaseManager(granularity)
        return self._db_managers[granularity]
    
    def retrieve_historical_data(self, request: DataRetrievalRequest) -> DataRetrievalResult:
        """
        Retrieve historical data for a given symbol and date range.
        
        Args:
            request: DataRetrievalRequest with symbol, dates, and granularity
            
        Returns:
            DataRetrievalResult with retrieved data or error information
        """
        logger.info(f"Starting historical data retrieval for {request.symbol}", 
                   start_date=request.start_date, end_date=request.end_date)
        
        try:
            # Validate symbol
            if not SymbolValidator.is_valid_symbol(request.symbol):
                error_msg = f"Invalid symbol format: {request.symbol}"
                logger.error(error_msg)
                return DataRetrievalResult(
                    symbol=request.symbol,
                    success=False,
                    data_points=[],
                    error_message=error_msg
                )
            
            # Check if symbol is available
            if not self.client.is_symbol_available(request.symbol):
                error_msg = f"Symbol {request.symbol} is not available for trading"
                logger.error(error_msg)
                return DataRetrievalResult(
                    symbol=request.symbol,
                    success=False,
                    data_points=[],
                    error_message=error_msg
                )
            
            # Retrieve data from API
            raw_data = self._fetch_data_from_api(request)
            
            if not raw_data:
                error_msg = f"No data retrieved for {request.symbol}"
                logger.warning(error_msg)
                return DataRetrievalResult(
                    symbol=request.symbol,
                    success=False,
                    data_points=[],
                    error_message=error_msg
                )
            
            # Transform and validate data
            data_points = self._transform_api_data(raw_data, request.symbol)
            
            logger.info(f"Successfully retrieved {len(data_points)} data points for {request.symbol}")
            
            return DataRetrievalResult(
                symbol=request.symbol,
                success=True,
                data_points=data_points
            )
            
        except Exception as e:
            error_msg = f"Failed to retrieve historical data for {request.symbol}: {str(e)}"
            logger.error(error_msg, exc_info=True)
            return DataRetrievalResult(
                symbol=request.symbol,
                success=False,
                data_points=[],
                error_message=error_msg
            )
    
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=2, max=10),
        retry=retry_if_exception_type((ConnectionError, TimeoutError, Exception))
    )
    def _fetch_data_from_api(self, request: DataRetrievalRequest) -> Optional[List[Dict[str, Any]]]:
        """
        Fetch raw data from Coinbase API with retry logic.
        
        Args:
            request: DataRetrievalRequest object
            
        Returns:
            List of raw data dictionaries or None if failed
        """
        try:
            # Convert datetime objects to Unix timestamp strings
            start_time = str(int(request.start_date.timestamp()))
            end_time = str(int(request.end_date.timestamp()))
            
            # Convert granularity from seconds to Coinbase API format
            granularity_map = {
                60: 'ONE_MINUTE',
                300: 'FIVE_MINUTE', 
                900: 'FIFTEEN_MINUTE',
                3600: 'ONE_HOUR',
                21600: 'SIX_HOUR',
                86400: 'ONE_DAY'
            }
            
            granularity_str = granularity_map.get(request.granularity, 'ONE_HOUR')
            
            logger.debug(f"Fetching data from API", 
                        symbol=request.symbol, 
                        start_time=start_time, 
                        end_time=end_time,
                        granularity=granularity_str)
            
            # Call Coinbase API for historical data using the new method
            historical_data = self.client.get_historical_candles(
                symbol=request.symbol,
                start=start_time,
                end=end_time,
                granularity=granularity_str
            )
            
            if historical_data and isinstance(historical_data, list):
                logger.debug(f"Retrieved {len(historical_data)} raw data points from API")
                return historical_data
            else:
                logger.warning("API returned empty or invalid data")
                return None
                
        except Exception as e:
            logger.error(f"API call failed: {e}")
            raise
    
    def _transform_api_data(self, raw_data: List[Dict[str, Any]], symbol: str) -> List[CryptoPriceData]:
        """
        Transform raw API data into CryptoPriceData objects.
        
        Args:
            raw_data: Raw data from Coinbase API
            symbol: Cryptocurrency symbol
            
        Returns:
            List of CryptoPriceData objects
        """
        data_points = []
        
        for item in raw_data:
            try:
                # Parse timestamp (Unix timestamp string)
                timestamp = datetime.fromtimestamp(int(item['start']))
                
                # Parse price data from dictionary format
                low_price = Decimal(str(item['low']))
                high_price = Decimal(str(item['high']))
                open_price = Decimal(str(item['open']))
                close_price = Decimal(str(item['close']))
                volume = Decimal(str(item['volume']))
                
                # Create data point
                data_point = CryptoPriceData(
                    symbol=symbol,
                    timestamp=timestamp,
                    open_price=open_price,
                    high_price=high_price,
                    low_price=low_price,
                    close_price=close_price,
                    volume=volume
                )
                
                data_points.append(data_point)
                
            except (ValueError, KeyError, TypeError) as e:
                logger.warning(f"Failed to parse data point: {item}, error: {e}")
                continue
        
        logger.debug(f"Transformed {len(data_points)} valid data points from {len(raw_data)} raw points")
        return data_points
    
    def retrieve_data_for_date_range(self, symbol: str, start_date: datetime, 
                                   end_date: datetime, granularity: int = 3600) -> DataRetrievalResult:
        """
        Convenience method to retrieve data for a date range.
        
        Args:
            symbol: Cryptocurrency symbol
            start_date: Start date for data retrieval
            end_date: End date for data retrieval
            granularity: Data granularity in seconds (default: 3600 = 1 hour)
            
        Returns:
            DataRetrievalResult object
        """
        request = DataRetrievalRequest(
            symbol=symbol,
            start_date=start_date,
            end_date=end_date,
            granularity=granularity
        )
        
        return self.retrieve_historical_data(request)
    
    def retrieve_recent_data(self, symbol: str, days: int = 7, granularity: int = 3600) -> DataRetrievalResult:
        """
        Retrieve recent data for a symbol.
        
        Args:
            symbol: Cryptocurrency symbol
            days: Number of days to retrieve (default: 7)
            granularity: Data granularity in seconds (default: 3600 = 1 hour)
            
        Returns:
            DataRetrievalResult object
        """
        end_date = datetime.utcnow()
        start_date = end_date - timedelta(days=days)
        
        return self.retrieve_data_for_date_range(symbol, start_date, end_date, granularity)
    
    def get_available_granularities(self) -> List[int]:
        """
        Get list of available granularities for data retrieval.
        
        Returns:
            List of granularity values in seconds
        """
        return [60, 300, 900, 3600, 21600, 86400]  # 1min, 5min, 15min, 1hr, 6hr, 1day
    
    def validate_date_range(self, start_date: datetime, end_date: datetime, granularity: int) -> bool:
        """
        Validate if date range is appropriate for given granularity.
        
        Args:
            start_date: Start date
            end_date: End date
            granularity: Granularity in seconds
            
        Returns:
            True if valid, False otherwise
        """
        duration = (end_date - start_date).total_seconds()
        max_duration = granularity * 300  # Coinbase limit: 300 data points per request
        
        return duration <= max_duration
    
    def retrieve_all_historical_data(self, symbol: str, granularity: int = 3600, 
                                   max_years_back: int = 5) -> DataRetrievalResult:
        """
        Retrieve all available historical data for a symbol by chunking requests.
        
        Args:
            symbol: Cryptocurrency symbol
            granularity: Data granularity in seconds (default: 3600 = 1 hour)
            max_years_back: Maximum years to go back (default: 5)
            
        Returns:
            DataRetrievalResult with all retrieved data
        """
        logger.info(f"Starting complete historical data retrieval for {symbol}", 
                   granularity=granularity, max_years_back=max_years_back)
        
        try:
            # Validate symbol
            if not SymbolValidator.is_valid_symbol(symbol):
                error_msg = f"Invalid symbol format: {symbol}"
                logger.error(error_msg)
                return DataRetrievalResult(
                    symbol=symbol,
                    success=False,
                    data_points=[],
                    error_message=error_msg
                )
            
            # Check if symbol is available
            if not self.client.is_symbol_available(symbol):
                error_msg = f"Symbol {symbol} is not available for trading"
                logger.error(error_msg)
                return DataRetrievalResult(
                    symbol=symbol,
                    success=False,
                    data_points=[],
                    error_message=error_msg
                )
            
            # Calculate date range
            end_date = datetime.utcnow()
            start_date = end_date - timedelta(days=max_years_back * 365)
            
            # Calculate chunk size based on granularity (max 300 data points per request)
            chunk_duration = granularity * 300
            chunk_timedelta = timedelta(seconds=chunk_duration)
            
            all_data_points = []
            current_start = start_date
            chunk_count = 0
            
            logger.info(f"Retrieving data in chunks of {chunk_duration} seconds", 
                       total_duration=(end_date - start_date).total_seconds())
            
            while current_start < end_date:
                chunk_count += 1
                current_end = min(current_start + chunk_timedelta, end_date)
                
                logger.debug(f"Processing chunk {chunk_count}: {current_start} to {current_end}")
                
                # Create request for this chunk
                chunk_request = DataRetrievalRequest(
                    symbol=symbol,
                    start_date=current_start,
                    end_date=current_end,
                    granularity=granularity
                )
                
                # Retrieve data for this chunk
                chunk_result = self.retrieve_historical_data(chunk_request)
                
                if not chunk_result.success:
                    logger.warning(f"Chunk {chunk_count} failed: {chunk_result.error_message}")
                    # Continue with next chunk instead of failing completely
                    current_start = current_end
                    continue
                
                if chunk_result.data_points:
                    all_data_points.extend(chunk_result.data_points)
                    logger.debug(f"Chunk {chunk_count} retrieved {len(chunk_result.data_points)} data points")
                
                # Move to next chunk
                current_start = current_end
                
                # Add small delay to respect rate limits
                import time
                time.sleep(0.1)
            
            logger.info(f"Complete historical data retrieval finished", 
                       total_chunks=chunk_count, total_data_points=len(all_data_points))
            
            return DataRetrievalResult(
                symbol=symbol,
                success=True,
                data_points=all_data_points
            )
            
        except Exception as e:
            error_msg = f"Failed to retrieve complete historical data for {symbol}: {str(e)}"
            logger.error(error_msg, exc_info=True)
            return DataRetrievalResult(
                symbol=symbol,
                success=False,
                data_points=[],
                error_message=error_msg
            )


# Global historical data retriever instance
data_retriever = HistoricalDataRetriever()

File: cpp-cohida/src/data/DataRetriever.cpp
#include "data/DataRetriever.h"
#include <iomanip>

std::string format_time_point(const std::chrono::system_clock::time_point& tp) {
    auto t = std::chrono::system_clock::to_time_t(tp);
    std::tm tm;
    gmtime_r(&t, &tm);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%d %H:%M:%S");
    return oss.str();
}

namespace data {

DataRetriever::DataRetriever() : is_retrieving_(false) {
    coinbase_client_ = std::make_unique<api::CoinbaseClient>();
    
    if (!coinbase_client_->is_authenticated()) {
        LOG_WARN("Coinbase client not authenticated");
    }
    
    LOG_INFO("DataRetriever initialized successfully");
}

DataRetriever::~DataRetriever() {
    LOG_INFO("DataRetriever shutting down");
}

DataRetrievalResult DataRetriever::retrieve_historical_data(const std::string& symbol,
                                                           system_clock::time_point start_date,
                                                           system_clock::time_point end_date,
                                                           int granularity) {
    DataRetrievalRequest request(symbol, start_date, end_date, granularity);
    return retrieve_historical_data(request);
}

DataRetrievalResult DataRetriever::retrieve_historical_data(const DataRetrievalRequest& request) {
    is_retrieving_ = true;
    LOG_INFO("Starting historical data retrieval for " + request.symbol);

    try {
        auto data_points = _fetch_data_from_api(request);
        
        if (data_points.empty()) {
            LOG_WARN("No data points retrieved for symbol: " + request.symbol);
            is_retrieving_ = false;
            return DataRetrievalResult(request.symbol, true, {}, "No data points available");
        }
        
        LOG_INFO("Successfully retrieved " + std::to_string(data_points.size()) + 
                  " data points for symbol: " + request.symbol);
        
        is_retrieving_ = false;
        return DataRetrievalResult(request.symbol, true, data_points);
    } catch (const std::exception& ex) {
        LOG_ERROR("Error retrieving historical data: " + std::string(ex.what()));
        is_retrieving_ = false;
        return DataRetrievalResult(request.symbol, false, {}, ex.what());
    }
}

std::vector<models::CryptoPriceData> DataRetriever::_fetch_data_from_api(const DataRetrievalRequest& request) {
    try {
        auto candles = coinbase_client_->get_historical_candles(
            request.symbol,
            request.start_date,
            request.end_date,
            request.granularity
        );
        
        LOG_DEBUG("API returned " + std::to_string(candles.size()) + " candles for symbol: " + request.symbol);
        return candles;
    } catch (const api::ApiException& ex) {
        LOG_ERROR("API Error: " + std::string(ex.what()));
        throw;
    } catch (const std::exception& ex) {
        LOG_ERROR("Error fetching data from API: " + std::string(ex.what()));
        throw;
    }
}

std::vector<models::CryptoPriceData> DataRetriever::_transform_api_data(const std::vector<json>& raw_data,
                                                                      const std::string& symbol) {
    std::vector<models::CryptoPriceData> data_points;
    data_points.reserve(raw_data.size());

    for (const auto& candle : raw_data) {
        try {
            if (!candle.is_array() || candle.size() < 6) {
                LOG_WARN("Invalid candle format: " + candle.dump());
                continue;
            }

            auto timestamp = system_clock::time_point(seconds(static_cast<long long>(candle[0])));
            
            models::CryptoPriceData data_point(
                symbol,
                timestamp,
                models::Decimal(candle[3].get<std::string>()),
                models::Decimal(candle[2].get<std::string>()),
                models::Decimal(candle[1].get<std::string>()),
                models::Decimal(candle[4].get<std::string>()),
                models::Decimal(candle[5].get<std::string>())
            );

            data_points.push_back(std::move(data_point));
        } catch (const std::exception& ex) {
            LOG_WARN("Failed to transform candle data: " + std::string(ex.what()));
            continue;
        }
    }

    LOG_DEBUG("Transformed " + std::to_string(data_points.size()) + " data points from API response");
    return data_points;
}

system_clock::time_point DataRetriever::_find_earliest_available_data(const std::string& symbol, int granularity,
                                                                system_clock::time_point max_test_date) {
    LOG_INFO("Finding earliest available data for " + symbol);
    
    const int MAX_YEARS_BACK = 10;
    const int TEST_WINDOW_DAYS = 7;
    
    auto current = system_clock::now();
    for (int years_back = 1; years_back <= MAX_YEARS_BACK; ++years_back) {
        auto test_start = current - years(years_back);
        auto test_end = test_start + days(TEST_WINDOW_DAYS);
        
        if (test_end > max_test_date) {
            LOG_DEBUG("Skipping test for " + std::to_string(years_back) + " years back - exceeds max date");
            continue;
        }
        
        LOG_DEBUG("Testing data availability from " + 
                  format_time_point(test_start) + " to " + 
                  format_time_point(test_end));
        
        try {
            DataRetrievalRequest request(symbol, test_start, test_end, granularity, true);
            auto test_result = retrieve_historical_data(request);
            
            if (test_result.success && !test_result.data_points.empty()) {
                auto min_timestamp = std::min_element(
                    test_result.data_points.begin(),
                    test_result.data_points.end(),
                    [](const models::CryptoPriceData& a, const models::CryptoPriceData& b) {
                        return a.timestamp < b.timestamp;
                    })->timestamp;
                
                LOG_DEBUG("Found data from " + std::to_string(years_back) + 
                          " years back: " + format_time_point(min_timestamp));
                return min_timestamp;
            }
        } catch (const std::exception& ex) {
            LOG_DEBUG("No data found from " + std::to_string(years_back) + " years back");
        }
    }
    
    LOG_WARN("Failed to find earliest available data, using default 1 year back");
    return current - years(1);
}

DataRetrievalResult DataRetriever::retrieve_all_historical_data(const std::string& symbol, int granularity,
                                                               std::optional<int> max_records) {
    LOG_INFO("Starting complete historical data retrieval for " + symbol);
    
    try {
        auto end_date = system_clock::now();
        auto start_date = _find_earliest_available_data(symbol, granularity, end_date - years(10));
        
        LOG_INFO("Auto-detected earliest data for " + symbol + ": " + 
                  format_time_point(start_date));
        
        std::vector<models::CryptoPriceData> all_data_points;
        
        const int MAX_CANDLES_PER_REQUEST = 200;
        const int REQUEST_INTERVAL_SECONDS = granularity * MAX_CANDLES_PER_REQUEST;
        
        auto chunk_start = start_date;
        int chunk_count = 0;
        
        while (chunk_start < end_date) {
            auto chunk_end = chunk_start + seconds(REQUEST_INTERVAL_SECONDS);
            if (chunk_end > end_date) {
                chunk_end = end_date;
            }
            
            LOG_INFO("Processing chunk " + std::to_string(chunk_count + 1) + ": " +
                      format_time_point(chunk_start) + " to " + format_time_point(chunk_end));
            
            try {
                DataRetrievalRequest request(symbol, chunk_start, chunk_end, granularity, true);
                auto chunk_result = retrieve_historical_data(request);
                
                if (chunk_result.success && !chunk_result.data_points.empty()) {
                    all_data_points.insert(
                        all_data_points.end(),
                        std::make_move_iterator(chunk_result.data_points.begin()),
                        std::make_move_iterator(chunk_result.data_points.end())
                    );
                    
                    LOG_DEBUG("Chunk " + std::to_string(chunk_count + 1) + 
                              " retrieved " + std::to_string(chunk_result.data_points.size()) + " data points");
                } else {
                    LOG_WARN("Chunk " + std::to_string(chunk_count + 1) + " failed: " + chunk_result.error_message);
                }
            } catch (const std::exception& ex) {
                LOG_WARN("Chunk " + std::to_string(chunk_count + 1) + " failed: " + ex.what());
            }
            
            chunk_start = chunk_end + seconds(1);
            chunk_count++;
            
            if (max_records && static_cast<int>(all_data_points.size()) >= *max_records) {
                LOG_INFO("Reached maximum record limit: " + std::to_string(*max_records));
                all_data_points.erase(all_data_points.begin() + *max_records, all_data_points.end());
                break;
            }
        }
        
        LOG_INFO("Complete historical data retrieval finished");
        
        return DataRetrievalResult(symbol, true, all_data_points);
    } catch (const std::exception& ex) {
        LOG_ERROR("Error in complete historical data retrieval: " + std::string(ex.what()));
        return DataRetrievalResult(symbol, false, {}, ex.what());
    }
}

bool DataRetriever::validate_symbol(const std::string& symbol) {
    try {
        auto symbol_info = coinbase_client_->get_symbol_info(symbol);
        return symbol_info.has_value();
    } catch (const std::exception& ex) {
        LOG_ERROR("Error validating symbol: " + std::string(ex.what()));
        return false;
    }
}

void DataRetriever::_log_retrieval_progress(int current, int total) {
    if (total > 0) {
        double progress = static_cast<double>(current) / total * 100;
        LOG_INFO("Data retrieval progress: " + std::to_string(static_cast<int>(progress)) + "%");
    }
}

} // namespace data